pyrit/attacks/multi_turn/tree_of_attacks.py

```diff
@@ -0,0 +1,1029 @@
+            best_node = context.nodes[0]  # Assumes already sorted by score
+            context.best_conversation_id = best_node.objective_target_conversation_id
+            context.best_objective_score = best_node.objective_score
+
+    def _create_attack_node(self, *, context: TAPAttackContext, parent_id: Optional[str] = None) -> TreeOfAttacksNode:
+        """
+        Create a new attack node with the configured settings.
+
+        Constructs a TreeOfAttacksNode instance with all necessary components including
+        targets, scorers, converters, and attack metadata. If on-topic checking is enabled,
+        creates a specialized scorer configured for the current objective.
+
+        Args:
+            context (TAPAttackContext): The attack context containing the objective and other configuration.
+            parent_id (Optional[str]): The ID of the parent node in the tree, if any. If None,
+                the node will be a root-level node.
+
+        Returns:
+            TreeOfAttacksNode: A new node configured for the TAP attack, ready to
+                generate adversarial prompts and evaluate responses.
+        """
+        return TreeOfAttacksNode(
+            objective_target=self._objective_target,
+            adversarial_chat=self._adversarial_chat,
+            adversarial_chat_seed_prompt=self._adversarial_chat_seed_prompt,
+            adversarial_chat_system_seed_prompt=self._adversarial_chat_system_seed_prompt,
+            adversarial_chat_prompt_template=self._adversarial_chat_prompt_template,
+            objective_scorer=self._objective_scorer,
+            on_topic_scorer=self._create_on_topic_scorer(context.objective),
+            request_converters=self._request_converters,
+            response_converters=self._response_converters,
+            auxiliary_scorers=self._auxiliary_scorers,
+            attack_id=self.get_identifier(),
+            memory_labels=context.memory_labels,
+            desired_response_prefix=self._desired_response_prefix,
+            parent_id=parent_id,
+            prompt_normalizer=self._prompt_normalizer,
+        )
+
+    def _normalize_score_to_float(self, score: Optional[Score]) -> float:
+        """
+        Normalize any score to a float value between 0.0 and 1.0.
+
+        Args:
+            score: The score to normalize, or None.
+
+        Returns:
+            Float value between 0.0 and 1.0. Returns 0.0 if score is None.
+        """
+        if not score:
+            return 0.0
+
+        score_value = score.get_value()
+        if isinstance(score_value, bool):
+            return 1.0 if score_value else 0.0
+        elif isinstance(score_value, (int, float)):
+            return float(score_value)
+        else:
+            self._logger.warning(f"Unexpected score value type: {type(score_value)} with value: {score_value}")
+            return 0.0
+
+    def _get_completed_nodes_sorted_by_score(self, nodes: List[TreeOfAttacksNode]) -> List[TreeOfAttacksNode]:
```

Comment:
```suggestion
    def _get_completed_nodes_sorted_by_score(self, nodes: List[TreeOfAttacksNode | None]) -> List[TreeOfAttacksNode]:
```
I think we allow None here, right? Otherwise, we don't need to check `if node` below

Reply:
We do not allow `None` here. Statically, when you declare `nodes` as `List[TreeOfAttacksNode]` , static-type checkers like `mypy` complain when you try to pass in a list that looks like this `[node1, None, ... ]`. However, that does not prevent a runtime list from not having `None` values. Imagine you read nodes from a file. That list could have None in it but since you are doing that at runtime, Python will not give you any warnings. 

The check in the function tries to prevent runtime None being passed in.